pageserver/src/tenant/timeline.rs

```diff
@@ -2296,27 +2295,31 @@ impl Timeline {
         }
 
         let mut layers = self.layers.write().unwrap();
-        let mut new_layer_paths = HashMap::with_capacity(new_layers.len());
+        let mut new_layer_names = HashMap::with_capacity(new_layers.len());
         for l in new_layers {
             let new_delta_path = l.path();
 
-            let metadata = new_delta_path.metadata()?;
-
-            if let Some(remote_client) = &self.remote_client {
-                remote_client.schedule_layer_file_upload(
-                    &l.filename(),
-                    &LayerFileMetadata::new(metadata.len()),
-                )?;
-            }
+            let metadata = new_delta_path.metadata()?; // XXX schedule upload for previous iterations
 
             // update the timeline's physical size
             self.metrics.current_physical_size_gauge.add(metadata.len());
 
-            new_layer_paths.insert(new_delta_path, LayerFileMetadata::new(metadata.len()));
+            new_layer_names.insert(l.filename(), LayerFileMetadata::new(metadata.len()));
             let x: Arc<dyn PersistentLayer + 'static> = Arc::new(l);
             layers.insert_historic(x);
         }
 
+        // Now that we've successfully scheduled all the layer uploads,
+        // point the IndexPart to them as well.
+        // If we crash before that is done, the pageserver startup procedure
+        // will re-schedule the uploads.
+        if let Some(remote_client) = &self.remote_client {
+            remote_client
+                .schedule_layer_file_uploads(&new_layer_names)
```

Comment:
We didn't previously schedule an index file upload here. We relied on the index upload that's part of the `schedule_layer_file_deletion` call later for that. ISTM that we now upload the index file twice.

Reply:
Correct.

Reply:
I'll sketch out a batch interface

Reply:
I think we should make scheduling infallible. Here's the exhaustive list of places where scheduling can fail:

- `ensure!(layer_metadata.file_size().is_some())` => turn into an assert. Should we ever fail the assert, then pageserver restart will reschedule uploads. If we want, we can use the type system to differentiate LayerFileMetadata with / without `file_size`.
- `upload_queue.latest_metadata.to_bytes()?;` AFAICT, this can only fail with an `Err()` if `bin_ser.rs` which is more or less `bincode` with more or less standard encoding options returns an error. Looking at the bincode impl, I think it generally fails only if the bincode options specify a size limit. But we don't specify a size limit. 
- `guard.initialized_mut()?` 
  - panic if we're in NotInitialized state
  - return OK() without doing anything if we're in state `Stopped` . Or maybe actually queue everything but simply don't launch the ops.

Opinions?

Reply:
No strong opinions but in general this is a worthy goal, perhaps not so easily attainable right now. Also, making operations infallible adds cost of change later, because now we are in a special part which cannot fail and many changes also introduce new failure cases...

> type system to differentiate `LayerFileMetadata` with/without `file_size`

This was attempted during the initial add of that `Option<u64>`, but it was before on-demand parts started to be merged, related: #2683. My vague memory is that the problem is that before `IndexPart` was unified to `RemoteIndex` early enough to force this unification on `Option<u64>`. More notes here: https://github.com/neondatabase/neon/pull/2582#discussion_r995906756 (but then, there was also `missing_layers` which has since been removed). I haven't looked at the latest versions like this, but it would sound like there definitely should be a way to do this.

Note that in future we are going to have to handle file size + hash upgrading so we can have 0, 1 or 2 of those.

> AFAICT, this can only fail with ...

Additionally all serde serializers can fail with the Serialize impl. In previous codebases I have asserted that a type successfully serializes after some light testing. This would be a fuzzable problem.

> `guard.initialized_mut()?`

Just occured to me that we (should) now have panic => issue pipeline via sentry. I don't think we have seen it in action with pageserver though.